### ISSUE TITLE
AArch64 assembler

### ISSUE THREAD

**Issue by jeffpanici75** [rtrussell/BBCSDL#4]
I understand you're looking for support implementing an AArch64 inline assembler, which I'd be happy to do.   Let me know how you'd like to proceed.

**Comment by rtrussell** [comment]
That would be very helpful.  At the moment the project has two assemblers: **bbasmb_arm_32.c** which is a naive ARM32 assembler (with currently no support for floating-point or SIMD instructions) and **bbasmb_x86_64.c** (a table-based x86-64 assembler with fairly good coverage of unprivileged instructions).  They should tell you all you need to know about the interface to the rest of BBC BASIC.  I'm guessing that an aarch64 assembler would be somewhere in between in terms of complexity.

**Comment by jeffpanici75** [comment]
Sounds good.  I'll follow the style present in the existing modules and rename the bbasmb_arm_64.TODO.c and put the new implementation there.  Are there are guidelines you'd like me to follow for the pull request?

**Comment by rtrussell** [comment]
> Are there are guidelines you'd like me to follow for the pull request?

Nothing that I'm aware of, but at the age of 68 I struggle to keep up with these new-fangled ways of working!

**Comment by ghost** [comment]
Hi, Jeff, just to let you know that I've been working on this for the last few days. I'm probably duplicating your work, but it's worth it to get a good understanding of the Aarch64 machine code. We've been communicating at the link below, but Richard suggested I should write something here, instead:

https://stardot.org.uk/forums/viewtopic.php?p=323516

**Comment by jgharston** [comment]
I've started work on an ARM64 assembler. Every now and then I stick the code at http://mdfs.net/Software/ARM/Assembler/

**Comment by jgharston** [comment]
I've been ploughing through arithmetic and logic, and have got to loads and stores, which was the stumbling block when I started looking at this, so before going further it's probably best to compare what we've each got to.

The thing that was the steep initial step was actually getting a full, complete, list of absolutely all instructions, and their binary encoding. Ok, two things, I'll get the comfy chair. Every other CPU I've written an assembler for it's been simple to find a summary for the entire instruction set that fits on one or two pages - the ARM32 fits on half a page of A5! I ploughed through the 8000-page ARM reference Richard linked to and got it down to about four pages. OpsSorted.txt at the link.

But the next problem is that a lot of instructions don't actually map to the binary instruction in any properly meaningful manner. So I had gone down a dead end of previous assemblers where, eg, b6-b3=operation, b2-b0=address mode, and all operations can use all address modes. I was stuck trying to build a binary representation of all the different LD and ST instructions, but different LD/ST have different address modes they can use. Even worse, LDR/STR allow different registers that none of the other LD/ST instructions use.

Effectively they are all different instructions, and it's much easier to code the assembler to treat them as different instructions, so that's the stage I've got to so far, albeit the reference codes are built up from a bitmap of LD/ST, data size, signed/unsigned, pair/unpaired.

So it would be valuable to compare where we've each got to before plough further in different directions.

**Comment by jeffpanici75** [comment]
Work has been keeping me extra busy and I haven't had a chance to start looking into this.  I don't want to hold up progress others are making on this front.  I'm excited to follow progress!

**Comment by ghost** [comment]
I had similar problems, and just went for a brute force approach and refined the way I dealt with each instruction as I went along. Anyway, here's my approach, it's about 5000 lines of code. I think it assembles every instruction, the question is if it does it properly!

There are instructions in there that will never be used by someone writing BASIC. I'd like to work out a way to automatically compare the output listing of the assembler with the output from a full-blown assembler. I've checked a few examples, obviously, but it's hardly been rigorous.

Do you know if there's a way to configure BASIC as a command-line program, so it's possible to take a dissassembly, turn it into a BASIC program, then compare the output with the original?

[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6586454/bbasmb_arm_64.c.zip)

**Comment by rtrussell** [comment]
> Do you know if there's a way to configure BASIC as a command-line program, so it's possible to take a dissassembly,
> turn it into a BASIC program, then compare the output with the original?

Yes, configuring BBC BASIC as a command-line assembler is relatively straightforward.  However it involves using the mode in which the 'program counter' and the 'code origin' are separated (achieved by setting bit 2 of OPT) and I don't know whether you've implemented that (yet).  I'll throw together a little program based on the 64-bit x86 edition that you can try.

**Comment by ghost** [comment]
Yes, there's definitely a problem with opt 7 and O%, I took a disassembly of some other code and hacked it into a BASIC program, but the jump and literal LDR locations were absolute and nowhere near the location of the allocated memory, so that reported many errors. I "solved" it by adding P% to the locations, because using O% just crashed BASIC.

That mod would be great. (I'm running BASIC on a PC, so I can't actually run the generated code.)

**Comment by ghost** [comment]
I used BASIC to write the assembled code to a file, then:

od -t x4 -Ax aarch64code.bbc | less
aarch64-none-elf-objcopy -I binary -O elf64-littleaarch64 -B aarch64  aarch64code.bbc aarch64code.o
aarch64-none-elf-objdump --disassemble-all aarch64code.o | less

Spoiler alert: there are bugs!

**Comment by ghost** [comment]
Literally the second instruction was wrong. Affecting or using SP with an ADD (and related instructions) has to use extended register or immediate, not shifted register (which should be obvious from the use of <Wd|WSP> as opposed to simply <Wd> in the documentation.

**Comment by ghost** [comment]
simply < Wd >

**Comment by rtrussell** [comment]
> I'll throw together a little program based on the 64-bit x86 edition that you can try.

Sorry I haven't done this yet, hopefully I may get around to it tomorrow.

**Comment by rtrussell** [comment]
Here you go.  This program assembles the source file specified on the command line.  Note that it doesn't work in the Windows console-mode edition of BBC BASIC; I am investigating that.  But it should be OK in the GUI versions and in the console-mode editions on other platforms.

```
      REM Assembles the file specified on the command line

      INSTALL @lib$ + "stringlib"
      sp% = FNinstrq(@cmd$, " ", 0)
      IF sp% srcfile$ = MID$(@cmd$, sp%+1) ELSE srcfile$ = @cmd$
      p%% = 0 : REM Code origin

      REM Find base filename:
      dot% = FN_instrr(srcfile$, ".", 0)
      dir% = FN_instrr(srcfile$, "/", 0)
      IF dir% = 0 dir% = FN_instrr(srcfile$, "\", 0)
      IF dot% > dir% base$ = LEFT$(srcfile$, dot%-1) ELSE base$ = srcfile$

      REM Open input file:
      srcfile% = OPENIN(srcfile$)
      IF srcfile% = 0 ERROR 0, "Couldn't open file " + srcfile$

      REM Create output files:
      binfile$ = base$ + ".bin"
      binfile% = OPENOUT(binfile$)
      IF binfile% = 0 ERROR 0, "Couldn't create file " + binfile$
      lstfile$ = base$ + ".lst"
      lstfile% = OPENOUT(lstfile$)
      IF lstfile% = 0 ERROR 0, "Couldn't create file " + lstfile$
      CLOSE #lstfile%
      tmpfile$ = @tmp$ + "assemble.tmp.bbc"

      REM Allocate memory for assembled code:
      DIM o%% 1000000 : l%% = o%% + 1000000

      REM Two pass assembly:
      FOR pass% = 12 TO 15 STEP 3
        PTR#srcfile% = 0
        lino% = 0
        tmpfile% = OPENOUT(tmpfile$)
        PROCputline(tmpfile%, 0, "[OPT " + STR$pass%)

        WHILE NOT EOF#srcfile%
          src$ = FNgetline(srcfile%)
          lino% += 1
          IF ASCsrc$ <> &25 PROCputline(tmpfile%, lino%, src$)
        ENDWHILE

        PROCputline(tmpfile%, lino%, "]")
        BPUT #tmpfile%, 0
        CLOSE #tmpfile%

        ]^L% = l%% : ]^O% = o%% : ]^P% = p%%
        OSCLI "spool """ + lstfile$ + """"
        ON ERROR LOCAL IF FALSE THEN
          CALL tmpfile$
        ELSE
          e$ = REPORT$
          I% = INSTR(e$, " in module")
          IF I% e$ = LEFT$(e$, I%-1)
          PRINT e$ + " at line "; ERL
        ENDIF : RESTORE ERROR
        IF ERR = 17 EXIT WHILE
        *SPOOL

      NEXT pass%
      CLOSE #srcfile%

      FOR C% = 0 TO ]^P% - p%% - 1
        BPUT#binfile%, o%%?C%
      NEXT
      CLOSE #binfile%

      PRINT "Assembly completed, output in " binfile$ ", listing in " lstfile$
      END

      DEF FNgetline(F%)
      LOCAL A$
      A$ = GET$#F%
      IF A$ = "" IF PTR#F%>1 THEN PTR#F%=PTR#F%-2:IF BGET#F%<>BGET#F% A$=GET$#F%
      = A$

      DEF PROCputline(F%, L%, s$)
      BPUT #F%, LEN(s$) + 4
      BPUT #F%, L% MOD 256
      BPUT #F%, L% DIV 256
      PRINT #F%, s$
      ENDPROC

      DEF FNinstrq(A$, B$, S%)
      LOCAL I%, Q%
      REPEAT
        I% = INSTR(A$, B$, S%)
        Q% = INSTR(A$, """", S%)
        IF Q%=0 OR I%<Q% THEN = I%
        S% = INSTR(A$, """", Q%+1)+1
      UNTIL S%=1
      = 0

```

**Comment by ghost** [comment]
Very nice! I only just spotted the console directory...

I get an error:
File or path not found at line 4
    4 INSTALL @lib$ + "stringlib"

Here's a slightly more readable version of what went before, with a couple of bugs fixed for good measure...
[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6593351/bbasmb_arm_64.c.zip)

In case it's not clear, generally (once I got the hang of it, at least), the I grouped mnemonics by related syntax and/or function (which usually implies encoding), the code identifies what variation of syntax is being used, checks the parameters for size and value, initialises the bit pattern in instruction according to the subset of mnemonics being assembled, finally filling in the parameters in their appropriate places.

I should probably revisit the various LDR/STR options, which were one of the first things I worked on, there's no doubt some more commonality there than I initially noticed.

For what it's worth, the code is packed with "magic numbers", but my justification for that is that if any of them changed, it wouldn't be the same processor we're assembling for. It also makes certain mistakes easier to spot; just look for duplicate patterns!

Which I've just done, and found several more mistakes! :(

**Comment by rtrussell** [comment]
> I get an error:
> File or path not found at line 4

That shouldn't happen if you've installed from the binary distribution, the zip file extracts all the files (including the libraries) to the correct relative places.   If you're building from source you need to make sure you run the binary in the project directory (note that the last line of the makefile is `cp bbcbasic ../../`).

**Comment by ghost** [comment]
So simple! Thanks, that's what I was missing. It works fine, now.

**Comment by ghost** [comment]
I was just pondering what a rigorous test would look like, and I realised: computers are fast, these days!

I just tried running **seq 0 $(( 0xffffffff ))**, and only takes just over a minute, even on this old PC; I can probably run every possible 32-bit value into a disassembler, put that into the assembler, and compare the output to the input, and it won't take more than an hour or so!

**Comment by rtrussell** [comment]
> I can probably run every possible 32-bit value into a disassembler

So long as the disassembler behaves sensibly when presented with a non-existent opcode, anyway.   That's one advantage of a CPU with a fixed instruction length; try generating every possible x86-64 instruction, which can be from one byte to 16 bytes in length!

**Comment by ghost** [comment]
Yes! Real life isn't as accommodating as I'd hoped, I wonder if the disassembler from objdump is usable, without having to write files and execute processes for every value...

Undefined instructions are easy, it writes, e.g.:
   0:	000251c1 	.inst	0x000251c1 ; undefined

**Comment by ghost** [comment]
Oh, look. There are an awful lot of floating point and SIMD instructions...

383 variations, in fact. :(

**Comment by rtrussell** [comment]
> There are an awful lot of floating point and SIMD instructions...

Indeed.  All I would ask is that any assembler is architecturally capable of being extended to handle them at some future point.  My current 32-bit ARM assembler doesn't support them, and they're not necessary for the profiling and debugging capabilities that are the principal application for assembly language (i.e. writing interrupt service routines, which you can't do in BASIC).  Do you happen to know whether any version of Acorn's ARM BASIC V assembler supports them?

**Comment by ghost** [comment]
ARM BBC BASIC V has the following, I guess QADD, etc. are FP?

>HELP [
Assembly language is contained in [] and assembled at P%. Labels follow '.'.
Syntax:
SWI|SVC|DBG|HVC|SMC|SMI[<cond>] <expr>
BFC[<cond>] <reg>,#<expr>,#<expr>
BFI|SBFX|UBFX[<cond>] <reg>,<reg>,#<expr>,#<expr>
USAT|SSAT[<cond>] <reg>,#<expr>,<shift>
USAT16|SSAT16[<cond>] <reg>,#<expr>,<reg>
UXTB|UXTB16|UXTH|UXTAB|UXTAB16|UXTAH|SXTB|SXTB16|SXTH[<cond>] <reg>,<shift>
BKPT|HLT|UDF <expr>
ADC|ADD|AND|BIC|EOR|ORR|RSB|RSC|SBC|SUB[<cond>][S] <reg>,<reg>,<shift>
MOV|MVN[<cond>][S] <reg>,<shift>
MOV[T|W][<cond> <reg>,#<expr>
CMN|CMP|TEQ|TST[<cond>][S|P] <reg>,<shift>
CLZ|RBIT|REV|REVSH|REV16[<cond>] <reg>,<reg>
CRC32[C]<B|H|W> <reg>,<reg>,<reg>
QADD[8|16]|QSUB[8|16]|USAD8|USADA8|QDADD|QDSUB[<cond>] <reg>,<reg>,<reg>
UADD|UHADD|UQADD|UQSUB|USUB|SADD|SHADD|SSUB|SHSUB<8|16>[<cond>] <reg>,<reg>,<reg>
QASX|QSAX|UQASX|UQSAX|SHASX|SHSAX|SSAX|SASX|USAX|UASX|SEL[<cond>] <reg>,<reg>,<reg>
MUL[<cond>][S] <reg>,<reg>,<reg>
MLA|MLS|UMULL|UMLAL|SMULL|SMLAL[<cond>][S] <reg>,<reg>,<reg>,<reg>
UMAAL[<cond>] <reg>,<reg>,<reg>,<reg>
SMUL<W|B|T><B|T>[<cond>] <reg>,<reg>,<reg>
SMLA[L]<W|B|T><B|T>[<cond>] <reg>,<reg>,<reg>,<reg>
SM<LA[L]|LS[L]|UA|US>D[X][<cond>] <reg>,<reg>,<reg>,<reg>
SMM<LA|LS|UL>[R][<cond>] <reg>,<reg>,<reg>,<reg>
LDR|STR[<cond>][B|T|BT|SB|SBT|H|HT|SH|SHT|D] <reg>, '[ <reg>[,<shift>] '] [,<shift>][!]
LDA|STL[<cond>][B|H] <reg>, '[ <reg> ']
LDREX|LDAEX|STREX|STLEX[B|H|D][<cond>] <reg>, '[ <reg> ']
LDM|STM[<cond>]DA|DB|EA|ED|FA|FD|IA|IB <reg>[!],{<reg list>}[^]
RFE<DA|DB|EA|ED|FA|FD|IA|IB> <reg>[!]
SRS<DA|DB|EA|ED|FA|FD|IA|IB> SP[!],#<expr>
SWP[<cond>][B] <reg>,<reg>, '[ <reg> ']
PLD[W]|PLI '[ <reg>[,<shift>] ']
PKH<BT|TB>[<cond>] <reg>,<reg>,<shift>
PUSH|POP[<cond>] <reg list>
B[L][<cond>] <label>
BLX <label>
BX|BLX|BXJ[<cond>] <reg>
SDIV|UDIV[<cond>] <reg>,<reg>,<reg>
WFE|WFI|SEV[L]|YIELD|ESB|CSDB[<cond>]
DMB|DSB|ISB [SY|SYST|ST|LD|<ISH|NSH|OSH>[ST|LD]]
TSB[<cond>] CSYNC
SETEND <BE|LE>
SETPAN #<expr>
MRC|MCR[<cond>|2] <copro>,<expr>,<reg>,<cpreg>,<cpreg> [,<expr>]
MCRR|MRRC[<cond>] <copro>,<expr>,<reg>,<reg>,<cpreg>
CDP[<cond>|2] <copro>,<expr>,<cpreg>,<cpreg>,<cpreg> [,<expr>]
LDC|STC[<cond>|2][L] <copro>,<cpreg>, '[ <reg>[,#<expr>] '] [,#<expr>|{expr}][!]
CPS<ID|IE> <iflags[,#<expr>]>
CPS #<expr>
CLREX|ERET|SSBB|PSSBB
MRS[<cond>] <reg>,<psr>
MSR[<cond>] <psr>_[c][x][s][f],<reg>|#<expr>
ADF|MUF|SUF|RSF|DVF|RDF|POW
   |RPW|RMF|FML|FDV|FRD|POL[<cond>]<prec>[<round>] <fpreg>,<fpreg>,<fpop>
MVF|MNF|ABS|RND|SQT|LOG|LGN|EXP
   |SIN|COS|TAN|ASN|ACS|ATN|URD|NRM[<cond>]<prec>[<round>] <fpreg>,<fpop>
FLT[<cond>]<prec>[<round>] <fpreg>,<reg>
FIX[<cond>][<round>] <reg>,<fpreg>
WFS|RFS|WFC|RFC[<cond>] <reg>
CMF|CNF[E][<cond>] <fpreg>,<fpop>
LDF|STF[<cond>]<prec> <fpreg>, '[ <reg>[,#<expr>] '] [,#<expr>][!]
LFM|SFM[<cond>] <fpreg>,<expr>, '[ <reg>[,#<expr>] '] [,#<expr>][!]
LFM|SFM[<cond>]EA|FD <fpreg>,<expr>, '[ <reg> '] [!]
DCF|EQUF<prec> <expr>
OPT|=|DCB|EQUB|DCW|EQUW|DCD|EQUD|EQUS <expr>
ADR[<cond>] <reg>,<label>
ALIGN|NOP
where <shift>=<reg>|#<expr>|<reg>,ASL|LSL|LSR|ASR|ROR <reg>|#<expr>|RRX
and <cond>=AL|CC|CS|EQ|GE|GT|HI|HS|LE|LS|LT|LO|MI|NE|NV|PL|VC|VS
and <reg>=R0 to 15 or SP or LR or PC or <expr>
and <copro>=CP0 to 15 or <expr>
and <cpreg>=C0 to 15 or <expr>
and <fpreg>=F0 to 7 or <expr>
and <fpop>=F0 to 7 or #<expr>, where <expr>=0,0.5,1,2,3,4,5 or 10
and <prec>=S|D|E|P
and <round>=P|M|Z
and <psr>=CPSR|SPSR
and <iflags>=A|I|F

**Comment by ghost** [comment]
I've hacked (really hacked, you don't want to know!) together a test which will take a few hours to run through every possible instruction, and came across an oddity: the disassembler from binutils' objdump ignores (probably correctly) bits that are shown as (1) in an instruction description, so I'll have to do some post-processing on the test output. As long as there's one OK result for a given assembly instruction, it's OK.

grep 'stlxrb.*w8, w19, \[x0\]' /media/simon/OS/Other/All.txt
08088013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08088413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08088813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08088C13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08089013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08089413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08089813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
08089C13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808A013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808A413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808A813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808AC13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808B013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808B413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808B813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808BC13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808C013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808C413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808C813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808CC13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808D013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808D413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808D813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808DC13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808E013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808E413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808E813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808EC13	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808F013	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808F413	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808F813	stlxrb	w8, w19, [x0]	->	0808FC13 FAILED
0808FC13	stlxrb	w8, w19, [x0]	->	0808FC13 OK

**Comment by jgharston** [comment]
On 04-06-2021 13:47, Simon-Willcocks wrote:
> Oh, look. There are an awful lot of floating point and SIMD
> instructions...
> 
> 383 variations, in fact. :(

Yes, that got me tearing my hair. Tootling along doing alu
instructions, all common addressing modes, make a start on
loads & stores, many common addressing modes.... ah! STR
and LDR can change to be completely different instructions,
but only after you've worked out what registers it is using.
Grrrr. On any other platform they would be eg FLDR and FSTR
or something.

I've made a start going through your (Simon's) code. I think
most of mine can be thrown away, but I think I've got some
optimisations.

I started with Richard's method of having a mnenionuc opcode
lookup table for the base code for each instruction that is
then modified by the specific instruction, so eg

case (various arithmetic):
instruction=opcode[menemoic]<<24;
...
if (immediate) instruction |= something

instead of
case blah: instruction=foo
case blah: instruction=foo
case blah: instruction=foo
case blah: instruction=foo
etc.

You can get most instructions into an 8-bit byte, where you can't
you add the extra bits in the instruction-specific code.

I also made the reg() routine return the register width in the
top two bits, so then it can be OR'd striaght into the instruction,
eg:
r=reg(); instruction |= (r & BIT31) | (r & 31); // Rd
comma();
r=reg(); instruction |= (r & BIT31) | ((r & 31)<<5); // Rn
etc.

It has the side effect that an instruction is "promoted" to the
biggest register used, eg:

add w1,w2,x3 is assembled as add x1,x2,x3

I hadn't yet got to the point of making a decision as to if this
should be allowed, or be an assembly error. For the latter you'd
collect the 'width' bits across the instruction and then fault
if they all don't match. I think the maximum number of registers
usable is four, so I think I'd set the top four bits from the
reg() routine and do something like:

r=reg(); width |= (r & BIT31); instruction |= (r & BIT31) | (r & 31); // 
Rd
comma();
r=reg(); width |= (r & BIT30); instruction |= (r & BIT31) | ((r & 
31)<<5); // Rn
comma();
r=reg(); width |= (r & BIT29); instruction |= (r & BIT31) | ((r & 
31)<<10); // Rm
...
if ((width == 0) || (width == (14<<28)) else error("Mixed register 
widths");
// %000xxx -> all 32-bit, %111xxx -> all 64-bit

-- 
J.G.Harston - ***@***.*** - mdfs.net/jgh

**Comment by ghost** [comment]
I made a similar change, but I used bits 5 and 6, so you can say if (32 == reg_size( r )...

The automated test has thrown up several problems, usually having to do with me not being able to count bits. It works through every possible opcode in about 2 hours, but the disassembler apparently knows instructions I don't (saddv, anyone?), and the output files are a bit excessive.

-rwxrwxrwx 1 simon simon 79G Jun  6 16:29 /media/simon/OS/Other/All.txt
-rwxrwxrwx 1 simon simon 43G Jun  6 16:29 /media/simon/OS/Other/Fail.txt
-rwxrwxrwx 1 simon simon 36G Jun  6 16:29 /media/simon/OS/Other/Pass.txt

I'm having difficulty attaching the latest version, I'll try again in a minute.

**Comment by ghost** [comment]
Half the instructions that have a bit to say it's working on a 64-bit register have them at bit 30!

(Upload still not working, sorry!)

**Comment by jeffpanici75** [comment]
Not sure this will be helpful to others but thought I'd mention it.  ARM releases their architecture specifications as executable code: https://developer.arm.com/-/media/developer/products/architecture/armv8-a-architecture/A64_v82A_ISA_xml_00bet3.1.tar.gz

If you unpack this tarball there's a folder with an XML file for every valid instruction encoding.  The XML includes ASL for both the decoding and execution of every instruction.  Note that ASL is like any other programming language: it has shared code (libraries) that are found in other directories of this archive.  Every aspect of the hardware is described this way.  ARM uses this for automated verification but we could use it to automate building an assembler/disassembler.  ARM has an example ASL interpreter: https://github.com/ARM-software/asl-interpreter

I'm sure there are other ASL interpreters written in more popular languages but I never bothered looking because I'm OK working with OCaml.

Anyhow, I was going to machine generate most (all?) of the assembler/disassembler this way.

**Comment by ghost** [comment]
Wow, I never expected something so powerful to be publically available! I might have a look at starting again and auto-generating the same kind of code...

On the other hand, testing is ongoing.

I worked out how to deal with encodings that don't match my encodings for the same assembly instruction: if the instruction I generate from the disassembly of the instruction we're looking it, I disassemble my attempt and compare the two assembly instructions. If they match, I count that as a "win" (labelled duplicate).

Apparently "adr xzr, address" is valid assembler and "adr sp, address" isn't. Strange.

Hey, uploads are working again!
[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6610749/bbasmb_arm_64.c.zip)

**Comment by ghost** [comment]
Does anyone know what on earth the difference is between:

ldrsb	wzr, [sp, xzr, sxtx #0]
and
ldrsb	wzr, [sp, xzr, sxtx]

(Other than the S bit has to be set in the former?)

"<amount> Is the index shift amount, it must be #0 , encoded in "S" as 0 if omitted, or as 1 if present."
But what difference does it make?

**Comment by rtrussell** [comment]
> But what difference does it make?

I'm guessing none.  Is there some reason why you think there is, or should be, a difference?

**Comment by ghost** [comment]
There's a bit to say if it's there or not!

**Comment by rtrussell** [comment]
> There's a bit to say if it's there or not!

Yes, but isn't this a case where the **S** bit is common across a number of different instructions, in most of which it **does** have an effect but in the case of this particular instruction variant it doesn't?  Normally **S** indicates an optional shift of the index value: a shift of 4 bits (*16) with a 128-bit load, 3 bits (*8) with a 64-bit load, 2 bits (*4) with a 32-bit load and 1 bit (*2) with a 16-bit load.  But when it's an 8-bit load, as in this case, there is no option to shift the index, even though the S bit is still present in the instruction encoding.  Hence whether you set it or not, the shift is zero.

In the specific case of this instruction an assembler might typically accept either no shift value, in which case it would not set the S bit, or a shift of #0 (that being the only allowed value), in which case it would set the S bit.  The effect would be identical, and setting the S bit indicative only of how the programmer coded the instruction.  But more generally it would accept either no shift value (setting S to zero) or the unique allowed value of #0, #1, #2, #3 or #4 as appropriate (setting S to one).

**Comment by ghost** [comment]
Yes, it looks like it "only" hurts my testing.

**Comment by ghost** [comment]
If anyone's still paying attention, the latest version, as far as I can tell, handles all the basic instructions correctly. What's left are the FP/SIMD instructions, and instructions introduced since the version of the DDI0487 (C) I was working with until recently. For some reason, I didn't expect more instructions to be added to a processor!

At least, of the 1,533,460,215 instructions objdump thinks are valid, none of them are assembled to the wrong instruction. But there are nearly 750 instruction classes that aren't assembled, yet.

It occurs to me that, perhaps, FP and SIMD instructions are important to BASIC programmers trying to extract more speed from their programs?
[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6650868/bbasmb_arm_64.c.zip)

I don't know if I can upload my test program, because it involves GNU code, and if I release it, it might infect BBCSDL. (And it's pretty ugly!) It runs through every possible instruction in about ninety minutes.

**Comment by rtrussell** [comment]
> It occurs to me that, perhaps, FP and SIMD instructions are important to BASIC programmers trying to extract more speed from their programs?

It's potentially the case, certainly, but since the 32-bit ARM assembler in BBCSDL doesn't currently support FP or SIMD instructions, anybody writing cross-platform code is going to be stuck anyway.

I'll try incorporating your code in an experimental build of BBCSDL; watch this space!

**Comment by rtrussell** [comment]
Preliminary results from trying to incorporate your code in the Console Edition build for the 'Apple Silicon' Mac (compiled using clang).  It is issuing four warnings, not necessarily serious but best eliminated (a couple look like they might be an unwise assumption about the signedness of 'char', which varies between platforms):
![Screenshot 2021-06-14 at 22 36 25](https://user-images.githubusercontent.com/42455052/121963465-0d62cd00-cd62-11eb-9b84-adc666123c39.png)

**Comment by ghost** [comment]
You're right, I didn't spot the first two because the tests all use lower case. Just change the declaration of code to unsigned char *code. The third is a hangover from earlier, when the value was inverted (a mistake), the line can simply by deleted. The rest are because the variable is not set in the default case of the switch, which never returns anyway, and will go away by simply initialising only_32bit to 0.

I will watch this space more carefully, and press refresh from time to time!
[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6656804/bbasmb_arm_64.c.zip)

**Comment by rtrussell** [comment]
>  I will watch this space more carefully, and press refresh from time to time!

Great, thanks.  Sadly your latest version is not compiling in clang, there are two warnings but more seriously it is reporting an implicit function declaration:
![Screenshot 2021-06-15 at 17 23 00](https://user-images.githubusercontent.com/42455052/122089933-2bcad600-cdff-11eb-9f24-b8deca3ef1bd.png)

**Comment by ghost** [comment]
Ooh, my mistake! I used to have a function for SIMD instructions, then I thought I might as well just put it in the massive switch statement with the others, and I forgot to get rid of the other calls. Just a minute.

**Comment by ghost** [comment]
How about this?
[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6657433/bbasmb_arm_64.c.zip)

**Comment by rtrussell** [comment]
> How about this?

That's much better.  I'm still getting the two warnings (see screen shot below) which can probably be fixed by casts, but they shouldn't be affecting operation.  I've noticed a couple of anomalies related to the displayed listing (compare the output from yours with the 64-bit x86 assembler, below):
1.  If I enter `[nop` nothing is displayed but if instead I enter `[nop:` (with a trailing colon) it is.   It looks like the listing is somehow being 'deferred' until the following instruction rather than being output immediately.
2.  There's no space between the address and the opcodes, possibly because you're not allowing for an address with more than 32 bits (I would have expected it to be padded with leading zeros to 64-bits).
![Screenshot 2021-06-15 at 22 35 05](https://user-images.githubusercontent.com/42455052/122127856-b9241f80-ce2b-11eb-977a-aaf50ed0968f.png)
![Untitled](https://user-images.githubusercontent.com/42455052/122128084-115b2180-ce2c-11eb-93b5-537573ba49b7.png)

**Comment by ghost** [comment]
The warnings can probably be avoided by removing the "unsigned", that's just me being dumb.

I don't think my stuff has a lot to do with displaying the opcodes, and when I try it on console/linux, it shows a space, but nothing with just "[nop", strangely. I'll have a look.

**Comment by ghost** [comment]
It's probably because I changed:

                        case 0x0D:                                    
                                newlin () ;                      
                                if (*esi == 0x0D)                               
                                        break ;                                   
                        case ':':                                       
                                if (liston & BIT4)                
                                    {
to:
            case 0x0D:                                                                
                newlin () ;                                                
                if (*esi == 0x0D || *esi == '\0')                  
                    break ;                                                         
            case ':':                                                               
                if (liston & BIT4)                                                            
                    {

I can't remember why I did that.

**Comment by ghost** [comment]
It works better without the change.

**Comment by rtrussell** [comment]
> The warnings can probably be avoided by removing the "unsigned"

Oh.  Isn't the "unsigned" the one you added to eliminate the previous 'comparison always fails' warnings (caused by 'char' being equivalent to 'signed char')?

> I don't think my stuff has a lot to do with displaying the opcodes

It's definitely behaving differently from the bbasmb_x86_64.c code from which you probably copied it.

> and when I try it on console/linux, it shows a space

Are you sure that's not simply because the allocated memory happen to be in the bottom 4 Gbytes and therefore needs only a 32-bit address?  You should be printing 16 hex digits.

**Comment by ghost** [comment]
"Oh. Isn't the "unsigned" the one you added to eliminate the previous 'comparison always fails' warnings (caused by 'char' being equivalent to 'signed char')?"
Yes, but I'm sometimes stupid! It's got to be cast to the same type it's being compared to.

I copied from bbasmb_arm_32.c.

The sprintfs still have a space after the oldpc, I don't know what's going on there.

**Comment by rtrussell** [comment]
> I copied from bbasmb_arm_32.c.
> The sprintfs still have a space after the oldpc, I don't know what's going on there.

Ah, if you copied the formatting of the listing from a 32-bit assembler it's bound to be wrong for a 64-bit assembler (twice as many hex digits in the address field!).  bbasmb_x86_64.c has:

`        sprintf (t, "%016llX ", (long long) oldpc) ;
`

I can make the change if you would prefer me to.

**Comment by ghost** [comment]
I made the changes in this version.
[bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6661772/bbasmb_arm_64.c.zip)

**Comment by rtrussell** [comment]
> I made the changes in this version.
> [bbasmb_arm_64.c.zip](https://github.com/rtrussell/BBCSDL/files/6661772/bbasmb_arm_64.c.zip)

Thanks; no warnings now!  There were some consequential changes that you didn't make (moving the opcode, label and instruction columns right 8 places to compensate for the increased address width) but I've done them myself and it's all looking good.

To put it to the test properly (that is, actually assemble and run some AArch64 code!) I need to set myself up with a Raspberry Pi running a 64-bit OS.  It'll be the first time I've tried to build a 64-bit ARM edition of BBC BASIC using GCC so there could be some pitfalls, I'll report progress here.

**Comment by rtrussell** [comment]
> I need to set myself up with a Raspberry Pi running a 64-bit OS.

I now have my Raspberry Pi running the 64-bit OS, but as anticipated I'm having trouble getting a 64-bit ARM build of BBCSDL to work (unrelated to the assembler).  Since it runs fine in iOS and on an 'Apple Silicon' Mac I'm guessing that this is a GCC versus Clang thing.  Bear with me.

**Comment by rtrussell** [comment]
What am I doing wrong here?  Adding a trailing ! is resulting in an 'Unknown instruction' error, whilst the online assembler I am using for comparison accepts it.  This is what I am using (with the listing columns adjusted): [bbasmb_arm_64.zip](https://github.com/rtrussell/BBCSDL/files/6665353/bbasmb_arm_64.zip)

![scrot](https://user-images.githubusercontent.com/42455052/122280073-b2a3af80-cee0-11eb-9a79-d98ff583944c.png)

**Comment by ghost** [comment]
OK, the first is correct, it translates to "stur    x30, [sp, #-16]"
Give me a minute...

**Comment by ghost** [comment]
My test program shows:
F81F0FFE        str     x30, [sp, #-16]!        ->      F81F0FFE OK

But it only tested one instruction at a time.
At or about line 598, put in a esi++. The "Unknown instruction" was the !.

**Comment by rtrussell** [comment]
> At or about line 598, put in a esi++

Yes, thanks, that fixed it.  On a related point, shouldn't I be able to use '**lr**' as a synonym for '**x30**'?  That's what it says [here](https://developer.arm.com/documentation/dui0801/a/BABBDBAD) but when I make that substitution in the instruction I listed previously I get a 'Bad register' error.

**Comment by ghost** [comment]
I guess that would be a yes. I just didn't think of it.

    if (strnicmp ((const char *)esi, "LR", 2) == 0) {
        esi += 2;
        return 30 | 64;
    }

Near the equivalent for SP in the reg function should fix that. "Link register. This is a synonym for X30.", so it's always 64 bits.

**Comment by rtrussell** [comment]
> I guess that would be a yes. I just didn't think of it.

Thanks, added.  In the comments it says "Registers can be 64-bit Xnn, XZR, SP, **LR**, or 32-bit Wnn, WZR" so it looks like you _did_ think of it, but didn't code it!  This program assembles and runs on the Raspberry Pi (if you can see any way to simplify or shorten it I would be interested): 

![scrot](https://user-images.githubusercontent.com/42455052/122369945-8f1b4c00-cf56-11eb-9058-f183705fa9d5.png)

**Comment by ghost** [comment]
There's a group of useful instructions CBZ, CBNZ, TBZ, TBNZ, which both test a register and jump.

CB[N]Z compares the register with zero, so CBZ W0, exit would replace the CMP and B.EQ.

The others test a bit in a register and jump if it's zero/non-zero, like: TBNZ W7, #12, code_for_when_bit_12_is_set

I wonder could you make simply "BL oswrch" work? It would somehow have to be moved to within 128MB of the location of the code.

**Comment by rtrussell** [comment]
> There's a group of useful instructions CBZ, CBNZ, TBZ, TBNZ, which both test a register and jump.

CBZ works, thanks.  You can see why some people think the elegance of 'classic' (32-bit) ARM has been lost, because you could conditionally restore multiple saved registers and return from the subroutine in one instruction: `popeq {r8-r9,pc} `

> I wonder could you make simply "BL oswrch" work? It would somehow have to be moved to within 128MB of the location of the code.

That's the rub.  There's no way to have any control over the relative displacement of BASIC's heap (which is where the program gets assembled to) and the code of the interpreter (which is where **oswrch** is) so you have to work on the basis that it could be anything.

**Comment by ghost** [comment]
Could you have a little table in the heap that does the indirect call part?

Assuming your code follows the Procedure Call Standard for the ARM 64-bit Architecture, registers 16 and 17 can be used for that kind of thing.

.oswrch
    ldr x16, actual_oswrch_location
    br x16

.actual_oswrch_location equq "oswrch"

Of course it would be ever so slightly less quick than having the register loaded with the final address, but it also de-clutters your code (and the PCS slows things down considerably, anyway, by not ensuring that most registers are unaffected, like with RISC OS SWIs).

For a more RISC OS-like experience, the indirect call could save and restore the registers that might be corrupted on the stack for you.

**Comment by rtrussell** [comment]
> Could you have a little table in the heap that does the indirect call part?

Putting it "in the heap" would mean it being declared as part of the assembly language source code (at least, that's the neatest way of storing constant data in the heap, especially when there are alignment requirements), making it a messier solution than what I have now!  In any case, the heap can be up to 256 Mbytes in BBCSDL (512 Mbytes in BB4W) so that doesn't necessarily guarantee it is in-range.

> For a more RISC OS-like experience

I can't imagine why anybody would want a more RISC OS like experience (disclaimer: I am a dyed-in-the-wool Windows enthusiast and have never used RISC OS - nor do I ever want to)!

**Comment by ghost** [comment]
If I'd known that, I'd never have agreed to help you! :)

Did you get into BBC BASIC from the original Beeb, then? (Mine's sitting right behind me.)

**Comment by rtrussell** [comment]
> Did you get into BBC BASIC from the original Beeb, then?

It was only quite recently that I found some paperwork proving that I was working on BBC BASIC (Z80) even before the BBC Micro was released (mid-1981, pretty much exactly 40 years ago)!

**Comment by ghost** [comment]
My school had a 380Z at that time; we had to load BASIC from tape at the beginning of the day.

**Comment by rtrussell** [comment]
The assembler seems to be working well and I think we can call this issue effectively resolved.  However I propose to keep it open for two reasons: (1) it will be a while before I've translated the debugger's assembler code to AArch64, and it's possible I'll encounter a problem on the way, and (2) I don't want to discourage other developers from continuing work on theirs.

The approximate code sizes of the assemblers we currently have are **9K** for arm-32 (no FP/SIMD), **15K** for x86-32 (table-based), **33K** for x86-64 (table-based) and **86K** for arm-64.  Given the relative complexities of the instruction sets my gut feeling is that a table-based AArch64 assembler could probably be substantially smaller than Simon's.

**Comment by ghost** [comment]
I don't have a working 64-bit ARM system to compile it on at the moment, but 86k sounds a little big. Did you compile with optimization on? Compiling under x86 generates code size of about **56k** without, and **26k** with optimization. (Looking at the size of the .text, .data, and .rodata entries of bbasmb.o.)

Whatever solution you choose, I've enjoyed learning the intricacies of the machine code. It's certainly not as clean as the old 32-bit (and 26-bit) systems, but it's very powerful.

**Comment by rtrussell** [comment]
> Did you compile with optimization on?

I'm compiling with -Os (optimize for size), this is the relevant extract from the makefile:
```
bbasmb.o: bbasmb_arm_64.c BBC.h
	$(CXX) $(CXXFLAGS) -Os $< -o $@

```

**Comment by ghost** [comment]
If you're looking at the size of the whole object file, it's got a lot of overhead. I think the amount is related to the number of symbols it has to link.  If you run strip on it, that will remove the overhead (but render the file unlinkable).

Running objdump on the object file shows very similar size to the x86 build I was doing. The code is in .text, read only data (the mnemonics, etc.) in .rodata*, and variables in .data. About 40K, total (optimised -Os). Remarkably similar to the size of the x86 code for the same file. Unless I'm missing something.

bbasmb_arm_64.o:     file format elf64-littleaarch64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         00006238  0000000000000000  0000000000000000  00000040  2^2
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, CODE
  1 .data         000001a0  0000000000000000  0000000000000000  00006278  2^3
                  CONTENTS, ALLOC, LOAD, RELOC, DATA
  2 .bss          00000000  0000000000000000  0000000000000000  00006418  2^0
                  ALLOC
  3 .rodata       00002208  0000000000000000  0000000000000000  00006418  2^3
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
  4 .rodata.str1.1 000018f6  0000000000000000  0000000000000000  00008620  2^0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .comment      00000012  0000000000000000  0000000000000000  00009f16  2^0
                  CONTENTS, READONLY

Anyway. I can see tables working well for disassembling, but I don't think it's that easy when assembling. I guess some kind of state machine, with a word or two per mnemonic that can be interpreted according to what appears next in the instruction could work...

**Comment by rtrussell** [comment]
> If you're looking at the size of the whole object file, it's got a lot of overhead.

I know, but I'm comparing the **relative** sizes of the different assemblers.  Unless there's some reason to suspect that the overheads are proportionally greater for the AArch64 assembler than the others, that should still be a valid comparison.  I did notice that whilst most of your 'internal' subroutines are declared as static, not all are; is there a reason for that?

> Anyway. I can see tables working well for disassembling, but I don't think it's that easy when assembling.

The x86-32 and x86-64 assemblers are table-based.  That approach seems to work well, and the x86-64 assembler is pleasingly compact considering the complex instruction set.   You might like to look at the source (bbasmb_x86_64.c) to see how I have tackled it.

**Comment by ghost** [comment]
No good reason for not declaring them static, no. It could be that for "hash" and "open_square", I was copying "comma", but "validated_imm..." there's no reason.

In my defence, "align", from bbasmb_arm_32.c is also non-static. :)

I've taken a quick look at bbasmb_x86_64.c, and it looks very neat, but as far as I can see, it doesn't have to cope with things like imm-that-is-9-bits-signed, imm-that-is-12-bits-word-aligned-and-positive, imm-that-is-a-repeating-pattern-of-x1s-and-y0s-rotated-by-z-bits, imm-that-is-log2-register-d-size-divided-by-8, etc.. The instruction opcode varies depending on things like that. Although, now I've typed that, I realise that perhaps it could be done (although some constants probably fall into multiple categories).

I hope that, having gone through all those instructions, if there had been an obvious pattern to the opcodes, I'd have spotted it. The opcodes are limited to 32 bits, and that takes quite a few tricks to get to work well. Maybe someone will prove me wrong. It wouldn't upset me, and I'll be happy to help test!

**Comment by rtrussell** [comment]
I'm attempting to assemble this instruction but it's not being accepted:
` uxtw x1, w24
`

This appears to be a valid instruction: it's being generated by the GCC compiler, and the online assembler I've been using as a reference [accepts it](http://shell-storm.org/online/Online-Assembler-and-Disassembler/?inst=uxtw+x1%2C+w24&arch=arm64&as_format=inline#assembly) and encodes it as follows:
`"\x01\x7f\x40\xd3"
`

The likelihood therefore is that your assembler is expecting a different (perhaps alternative) syntax.  It would be nice if it accepted the above syntax, but meanwhile what is it expecting in order to generate &**D3407F01** ?

**Comment by ghost** [comment]
D3407F01	ubfx	x1, x24, #0, #32	->	D3407F01 OK

UXTB and UXTH seem to be instructions, but UXTW isn't. (ARM DDI 0487G.a)
Isn't it the same as MOV x1, w24?

**Comment by rtrussell** [comment]
> UXTB and UXTH seem to be instructions, but UXTW isn't. (ARM DDI 0487G.a)

Well, if you can zero-extend a byte to a 64-bit register and a half-word to a 64-bit register you certainly should also be able to zero-extend a 32-bit register!  And it seems that indeed you can, but for some reason the 'official' instruction is the gobbledegook **ubfx x1, x24, #0, #32**.  I can well see why GCC's assembler and the online assembler accept the more obvious and memorable mnemonic UXTW, and I think we should too.

> Isn't it the same as MOV x1, w24?

The online assembler doesn't accept that, which doesn't surprise me since it doesn't specify whether the 32-bit register is sign-extended or zero-extended in the 64-bit register.  I presume UXTW means 'unsigned extend' and that is borne out by the online assembler also accepting SXTW which I guess means 'signed extend'.

What would be involved in making the assembler accept UXTW and SXTW?

**Comment by ghost** [comment]
Not much. I even commented on it not being an instruction at the time.
    case UXTH: // No UXTW instruction, just extend the word register

Add to the mnemonics and enum in alphabetical order:
"UXTB", "UXTH", "UXTL", "UXTL2", "UXTW", "UZP1", "UZP2", "WFE", "WFI",

UXTB, UXTH, UXTL, UXTL2, UXTW, UZP1, UZP2, WFE, WFI,

Then, after the three "case UXTH:" lines, add:

                    case UXTW:

                            case UXTW: instruction = 0x53000000; break;

                            case SXTW:
                            case UXTW: c2 = 31; break;

And one new line in the error check:

                         || ( word_data && mnemonic == SXTW)
                         || ( word_data && mnemonic == UXTW)
                         || (!word_data && mnemonic == UXTB)

**Comment by ghost** [comment]
I think my comment intended to suggest that this approach would work:

     8B3843E1      ADD X1, XZR, W24, UXTW

The ubfx is neater, but doesn't give you the option of setting flags (with ADDS).

**Comment by rtrussell** [comment]
> I even commented on it not being an instruction at the time.
> case UXTH: // No UXTW instruction, just extend the word register

Hah!  Actually implementing it is a much nicer solution, restoring the 'symmetry' with SXTW.  Thanks for the modification details, which were very clear; I'll test it later.

**Comment by rtrussell** [comment]
The UXTW instruction now seems to work, thanks.  For your interest (I'm not expecting any changes) other differences between what GCC outputs and what your assembler accepts are:

1. Numeric constants are not preceded by #.  It seems that the # is more a convention than a syntactical requirement.
2. Conditional branch instructions don't include a dot, so for example GCC uses 'beq' rather than 'b.eq'.

With those changes the source code created by GCC runs after being assembled in BBC BASIC so that's all very encouraging.  I hope to release the 64-bit Raspberry Pi edition of BBCSDL in a couple of weeks time.

**Comment by ghost** [comment]
I think making # optional could cause problems with BASIC variables:

        add x1, x2, x3%

Making the dot in conditional jumps optional would be a minor change; either modify the if statement that looks for the '.', or simply include mnemonics for "B.EQ", "BEQ", "B.NE", etc. (in aphabetical order, mind!), adding the cases to case B/BL, and inserting a switch statement instead of the if statement. (The enum entries might be a bit ugly, e.g. B_NE, or BdotNE?)

Good luck with the rollout.

**Comment by rtrussell** [comment]
> I think making # optional could cause problems with BASIC variables:

It doesn't in the case of the 32-bit x86 assembler:
```
>DIM P% 1000
>ebx% = 1234
>[mov eax,ebx
09C21FEC 8B C3                          mov eax,ebx
>[mov eax,ebx%
09C21FEE B8 D2 04 00 00                 mov eax,ebx%
```
But in any case not using variables starting with a register name is probably good practice.

**Comment by rtrussell** [comment]
In general, register names cause problems for the cruncher, because it must be able to recognise what is a variable name (and crunch it) and what is a register name (and leave it alone).  The way it currently works is that the cruncher first scans the entire program **but skipping any assembler code** and makes a list of all the variables it finds.  Then it crunches the entire program **including the assembler code**, but only crunching the variable names it previously found.

This strategy will break if the program uses a variable name anywhere that happens to clash with a register name, even if it's not used in the assembler code (more likely with ARM's register names than the x86's).  But I don't know of a better method (_BBC BASIC for Windows_' cruncher actually knows about assembly language syntax, but that's too much to expect in a cross-platform product).

**Comment by rtrussell** [comment]
I've now released v1.23a, with the AArch64 assembler bundled with the 64-bit Raspberry Pi edition.  Arguably I should have included it with the iOS edition too, but since the assembler can't be used in any conventional fashion (iOS forbids 'arbitrary code execution') I've not done so to reduce bloat.  I'll reconsider that decision if anybody feels strongly about it.

**Comment by rtrussell** [comment]
Sorry, reopening this because I tried adding the assembler to the iOS edition but clang is complaining about two lines:

```
                if (result.mode == -1)
                if (mod == -1)
```
in both cases complaining that comparing an **enum** with -1 always fails.  How do I fix this please?

**Comment by Rob-McKay** [comment]
As the size of an enum in C is implementation defined and -1 as an int may have more bits than the enum, the recommended way is to add -1 to the enum.  You could probably get away with an explicit cast as that tells the compiler that you know best.

**Comment by ghost** [comment]
Yes, the result of lookup is an integer that will be an enum value or -1. Maybe store the result of the lookup function in an integer, check if it's -1, then set the enum variable if it's not?

    int found = lookup( modifiers, sizeof( modifiers )/sizeof( modifiers[0] ) ) ;
    if (found == -1) error( 16, NULL );
    mod = found;

Something like that?

**Comment by rtrussell** [comment]
> Something like that?

Rob's solution (adding -1 to the enum) is simpler and arguably more elegant - and I've already made the change! - but do you have an objection to it, for example on the grounds that it's not strictly part of the enumeration?

**Comment by ghost** [comment]
No objection, Your Honour!

You're right, it is more elegant.

**Comment by rtrussell** [comment]
> You're right, it is more elegant.

OK, I'll stick with that.  Closing again (for the last time, hopefully!).
